### ISSUE TITLE
custom function in the expression of energy function

### ISSUE THREAD

**Issue by lslumass** [openmm/openmm#4350]
Hi,

I'm trying to get the angle between two planes (without shared points) in the expression of customforce. Is there any possible to define a function in the python script and then use it in the expression of the energy function?

Thank you!

**Comment by peastman** [comment]
Which custom force are you using?  And how are the planes defined?  If you have a normal vector, you can just take a dot product and pass it to `acos()`.

**Comment by wayuer19** [comment]
I'm using CustomHbondforce. There are three donor atoms to define a plane, and another three acceptors to define the second plane. Now get the angle between them.

**Comment by craabreu** [comment]
The Python API of OpenMM's [Library Layer](http://docs.openmm.org/latest/api-python/library.html) is a wrapper of compiled C++ objects. The functionality you mention doesn't exist, but it would significantly increase the flexibility of custom OpenMM forces if it existed. As I understand, accepting a general Python function as input would require a non-trivial implementation effort in OpenMM.

Here is my proposal for a compromise solution that is still very flexible:

A custom function could be constructed from `OpenMM::Force` objects and evaluated at an `OpenMM::Platform`. These could be specified independently of the forces and platform used to simulate the system of interest. For example, we could create a Gaussian basis function of four variables from a `CustomCoumpoundBondForce`:

```py
import openmm
import pandas as pd

df = pd.read_csv("gbf_parameters.csv")  # height, sigma, center_a, center_b, center_c, center_d

function = "height * exp(-(a^2 + b^2 + c^2 + d^2)/2)"
function += "; a = (x1 - center_a)/sigma"
function += "; b = (y1 - center_b)/sigma"
function += "; c = (z1 - center_c)/sigma"
function += "; d = (x2 - center_d)/sigma"
gbf = openmm.CustomCompoundBondForce(2, function)
gbf.addPerBondParameter("height")
gbf.addPerBondParameter("sigma")
for letter in "abcd":
    gbf.addPerParticleParameter(f"center_{letter}")
for params in df.itertuples(index=False):
    gbf.addBond([0, 1], params)
```

Then, a new `OpenMM::CustomFunction` class could be implemented as a subclass of `Lepton::CustomFunction` and be constructed like this:

```py
platform = openmm.Platform.getPlatformByName('CUDA')
custom_function = openmm.CustomFunction(platform)
custom_function.addVariable("x", 0)
custom_function.addVariable("y", 0)
custom_function.addVariable("z", 0)
custom_function.addVariable("x", 1)
custom_function.addSummand(gbf)
```

Finally, we must implement a new method to add these functions to the OpenMM custom forces. For instance:

```py
force = openmm.CustomCVForce("bias(cv1, cv2, cv3, cv4)")
force.addCustomFunction("bias", custom_function)
force.addCollectiveVariable("cv1", ...)
force.addCollectiveVariable("cv2", ...)
force.addCollectiveVariable("cv3", ...)
force.addCollectiveVariable("cv4", ...)
```

Internally, the CustomFunction object in the example above would create a system with two particles, add the passed CustomCompoundBondForce, and build a context from that system and the passed platform. The `evaluate` and `evaluateDerivative` methods overriding their `Lepton::CustomFunction` counterparts would involve calling `context.getState(getPotentialEnergy=True)` and `context.getState(getForces=True)`, respectively. Requests for derivatives with `order >= 2` could raise an exception.

Being a `Lepton::CustomFunction` subclass, this new class would naturally fit in the [TabulatedFunctionWrapper](https://github.com/openmm/openmm/blob/4c107329a24f16afeffd2fe6bfad01a40e1699b9/platforms/reference/src/SimTKReference/ReferenceCustomCVForce.cpp#L39C31-L39C55) framework recently added to the code. It could be implemented with a relatively small effort and create new possibilities right away. For instance, it would allow Metadynamics simulations with more than three collective variables. Another possibility, which I am particularly interested in, is implementing the [OPES method](https://pubs.acs.org/doi/10.1021/acs.jpclett.0c00497) and its [OPES-explore](https://pubs.acs.org/doi/full/10.1021/acs.jctc.2c00152) variant. These methods require much fewer Gaussians than Metadynamics and are particularly appealing for multidimensional CV spaces.

@peastman, can you please let me know your opinion?

**Comment by peastman** [comment]
I'm not sure I completely understand.  It sounds like it would be really expensive to compute, since it would require multiple trips back and forth between different forces?  I also don't see how it would be used to address this problem, computing the angle between two planes each defined by three atoms.

If you really want the ability to define forces with arbitrary Python code, have you looked at the [PyTorch plugin](https://github.com/openmm/openmm-torch)?  Anything that can be compiled to TorchScript can be turned into a force.

This is closely related to #2566.  Perhaps the solution would be to add a pair of functions, `anglebetweenvectors(a1, a2, b1, b2)` and `anglebetweenplanes(a1, a2, a3, b1, b2, b3)`?  Or we could try to support vector operations like `dot()` and `cross()`, similar to what CustomIntegrator does?

**Comment by wayuer19** [comment]
Maybe just need to add the function of getting the position of atoms and vector operations as you mentioned, so that lots of things can be done through them, like ``vec(a1)`` to get the position of a1 in vector form, and ``vec(a1, b1)`` to get the vector of a1-->b1. 

and this is really related to [#2566](https://github.com/openmm/openmm/issues/2566).  might merge them together.

**Comment by peastman** [comment]
That would be the most general solution.  Also very complicated to implement!  The issue is that the custom forces just evaluate a scalar function.  That's sort of true for CustomIntegrator, but it evaluates the function three times for each particle to compute the three components of a vector (like position or velocity).  It's not hard to adapt that to get the same effect as a vector function.  But with custom forces, the result of computing the expression must be a scalar (the energy).

**Comment by craabreu** [comment]
> I'm not sure I completely understand. It sounds like it would be really expensive to compute, since it would require multiple trips back and forth between different forces?

It is only a `Lepton:CustomFunction` implemented so that when `evaluate` is called, its arguments are passed via `setPositions` to a tiny context whose system contains very few (typically 1 or 2) particles and a single force. Then, the potential energy is evaluated and returned. In the case of a complex function like a sum of many Gaussian kernels, the ability to evaluate the kernels in parallel on a GPU seems like a big advantage. The overhead of communicating back and forth with such a tiny context should be small compared to this gain.

> I also don't see how it would be used to address this problem, computing the angle between two planes each defined by three atoms.

Sorry. I didn't mean to propose a solution for this specific problem. I should have opened a new issue.

**Comment by craabreu** [comment]
@wayuer19, as @peastman mentioned, the angle between the two planes equals the angle between their normal vectors. Therefore, if you evaluate `S=[(a2-a1)x(a3-a1)].[(b2-b1)x(b3-b1)]`, you can divide it by a product of distances and take the `acos` of the result. But there is the [quadruple product formula](https://en.wikipedia.org/wiki/Quadruple_product), with which you can compute S using dot products only. As a dot product is the product of two distances and a cosine, all you need are distances and angles. Does that make sense?

**Comment by wayuer19** [comment]
great idea. It does make sense. Thank you.
